homeassistant/components/roon/media_player.py

```diff
@@ -188,19 +188,23 @@ def update_state(self):
         self._unique_id = self.player_data["dev_id"]
         self._zone_id = self.player_data["zone_id"]
         self._output_id = self.player_data["output_id"]
-        self._name = self.player_data["display_name"]
-        self._is_volume_muted = self.player_data["volume"]["is_muted"]
-        self._volume_step = convert(self.player_data["volume"]["step"], int, 0)
         self._shuffle = self.player_data["settings"]["shuffle"]
-
-        if self.player_data["volume"]["type"] == "db":
-            volume = (
-                convert(self.player_data["volume"]["value"], float, 0.0) / 80 * 100
-                + 100
-            )
-        else:
-            volume = convert(self.player_data["volume"]["value"], float, 0.0)
-        self._volume_level = convert(volume, int, 0) / 100
+        self._name = self.player_data["display_name"]
+        try:
```

Comment:
i would recommend making a helper such as _get_volume_properties that does all this logic and returns the values that you can then set here

Reply:
```python
def _parse_volume(volume_data):
    """Parse volume data to determine volume levels and mute state."""
    volume = {
        "level": 0,
        "step": 0,
        "muted": False,
    }

    try:
        volume.muted = volume_data["is_muted"]
        volume.step = convert(volume_data["step"], int, 0)

        if volume_data["type"] == "db":
            volume_level = (
                convert(volume_level["value"], float, 0.0) / 80 * 100
                + 100
            )
            else:
                volume_level = convert(volume_data["value"], float, 0.0)

         volume.level = convert(volume, int, 0) / 100
    except KeyError:
        volume.muted = False
        volume.step = 0
        volume.level = 0

    return volume
```

Reply:
@ctalkington Nice pattern! I've refactored both the `volume` and `now playing` code in the style you describe. I needed to change slightly from your suggestion - because the initial dictionary lookup needed to be inside the exception block.  

Also the final assignments look like Javascript - I don't *think* there is a way to make that work in Python - but let me know if I missed your intent - and happy to amend.

Reply:
hah sorry, i dabble in a mix of languages! that was my bad